File: shared/repunit-circles.ts
// Copyright (c) 2025-2026 Capomastro Holdings Ltd. (Canada)
// Patent(s) Pending — All Rights Reserved — Applied Physics Division
//
// Repunit Circle Library
//
// Base-3 repunits R(n) = (3^n - 1) / 2 define the natural cycle hierarchy
// of the ternary circle geometry. These are GEOMETRIC constructs measured
// in circle-days (364-day rotations). They are NOT calendar years.
//
// The calendar year = 365 = 364 + DOT (Day Out of Time).
// Repunit circles never include DOT — they are pure geometry.
// To convert to calendar duration, DOT insertions must be computed separately.
//
// INVARIANT: These values derive from the repunit formula and
// the ternary circle identity 111111₃ = 364. No tuning parameters.

// ============================================================
// CORE: Repunit computation
// ============================================================

/**
 * Compute the n-th base-3 repunit: R(n) = (3^n - 1) / 2.
 * R(n) is the n-digit base-3 number consisting entirely of 1s.
 *
 * R(1) = 1, R(2) = 4, R(3) = 13, R(4) = 40, R(5) = 121,
 * R(6) = 364, R(7) = 1093, R(8) = 3280, R(9) = 9841
 *
 * @param n - Repunit length (number of 1-digits in base 3). Must be ≥ 1.
 */
export function repunit(n: number): number {
  if (n < 1 || !Number.isInteger(n)) {
    throw new Error(`Repunit index must be a positive integer, got ${n}`);
  }
  return (Math.pow(3, n) - 1) / 2;
}

/**
 * Repunit factorization identity: R(2n) = R(n) × (3^n + 1).
 * Returns the two factors.
 *
 * @param n - Half-index: returns factors of R(2n).
 */
export function repunitFactorization(n: number): {
  r2n: number;
  rn: number;
  cofactor: number; // 3^n + 1
} {
  const rn = repunit(n);
  const cofactor = Math.pow(3, n) + 1;
  return {
    r2n: rn * cofactor,
    rn,
    cofactor,
  };
}

// ============================================================
// CIRCLE DEFINITIONS
// ============================================================

export interface RepunitCircle {
  /** Repunit index n (number of base-3 digits) */
  index: number;
  /** Repunit value R(n) = (3^n - 1) / 2 */
  value: number;
  /** Circle-days: pure geometric duration (no DOT) */
  circleDays: number;
  /** Human-readable label */
  label: string;
  /** Base-3 repunit string (all 1s) */
  base3: string;
  /** Is R(n) prime? */
  isPrime: boolean;
}

/**
 * The repunit circle hierarchy.
 *
 * IMPORTANT: circleDays is a geometric measure, not a calendar measure.
 * To get calendar days, use `circleDaysToCalendarDays()`.
 */
export const REPUNIT_CIRCLES: RepunitCircle[] = [
  { index: 1, value: 1,    circleDays: 1,    label: "Unit",            base3: "1",       isPrime: false },
  { index: 2, value: 4,    circleDays: 4,    label: "Tetrad",          base3: "11",      isPrime: false },
  { index: 3, value: 13,   circleDays: 13,   label: "Radian",          base3: "111",     isPrime: true  },
  { index: 4, value: 40,   circleDays: 40,   label: "Minor Circle",    base3: "1111",    isPrime: false },
  { index: 5, value: 121,  circleDays: 121,  label: "Quarter Circle",  base3: "11111",   isPrime: false }, // 121 = 11²
  { index: 6, value: 364,  circleDays: 364,  label: "Full Circle",     base3: "111111",  isPrime: false }, // 364 = 2² × 7 × 13
  { index: 7, value: 1093, circleDays: 1093, label: "Triple Circle",   base3: "1111111", isPrime: true  },
  { index: 8, value: 3280, circleDays: 3280, label: "Ennead Circle",   base3: "11111111", isPrime: false },
  { index: 9, value: 9841, circleDays: 9841, label: "Grand Circle",    base3: "111111111", isPrime: false },
];

/**
 * Get a repunit circle by index.
 */
export function getRepunitCircle(n: number): RepunitCircle | undefined {
  return REPUNIT_CIRCLES.find(c => c.index === n);
}

// ============================================================
// CALENDAR CONVERSION (DOT-aware)
// ============================================================

/**
 * Full ternary circle in circle-days (R₆ = 364 = 111111₃).
 * This is the geometric constant. The calendar year is 364 + 1 DOT = 365.
 */
export const FULL_CIRCLE_DAYS = 364;

/**
 * Convert circle-days to calendar days by inserting DOT boundaries.
 *
 * The calendar year is 365 = 364 circle-days + 1 DOT.
 * DOT (Day Out of Time) occurs at the end of each full 364-day circle rotation,
 * specifically on November 11 (the Fibonacci partition point: 8 moons before, 5 after).
 *
 * For a span of `circleDays` geometric days:
 *   - completedCircles = floor(circleDays / 364)
 *   - each completed circle inserts 1 DOT
 *   - calendarDays = circleDays + completedCircles
 *
 * @param circleDays - Duration in geometric circle-days (no DOT).
 * @returns Duration in calendar days (with DOT insertions).
 */
export function circleDaysToCalendarDays(circleDays: number): number {
  if (circleDays < 0) throw new Error("Circle-days cannot be negative");
  const completedCircles = Math.floor(circleDays / FULL_CIRCLE_DAYS);
  return circleDays + completedCircles;
}

/**
 * Convert calendar days to circle-days by removing DOT boundaries.
 *
 * @param calendarDays - Duration in calendar days (with DOTs).
 * @returns Duration in geometric circle-days (pure geometry).
 */
export function calendarDaysToCircleDays(calendarDays: number): number {
  if (calendarDays < 0) throw new Error("Calendar days cannot be negative");
  // Each calendar year is 365 days = 364 circle-days + 1 DOT.
  // completedYears = floor(calendarDays / 365)
  // remaining = calendarDays - completedYears * 365
  // circleDays = completedYears * 364 + min(remaining, 364)
  const completedYears = Math.floor(calendarDays / 365);
  const remaining = calendarDays - completedYears * 365;
  return completedYears * FULL_CIRCLE_DAYS + Math.min(remaining, FULL_CIRCLE_DAYS);
}

// ============================================================
// PLATFORM CONSTANTS (add to shared/constants.ts → PLATFORM)
// ============================================================

/**
 * Repunit circle constants for PLATFORM object.
 */
export const REPUNIT_PLATFORM_CONSTANTS = {
  /** R₃ = 13 = T₇ = 1 radian = 111₃ */
  REPUNIT_R3: 13,
  /** R₄ = 40 = 1111₃ */
  REPUNIT_R4: 40,
  /** R₅ = 121 = 11111₃ = 11² */
  REPUNIT_R5: 121,
  /** R₆ = 364 = 111111₃ = full ternary circle */
  REPUNIT_R6: 364,
  /** R₇ = 1093 = 1111111₃ (prime) — key rotation circle */
  REPUNIT_R7: 1093,
  /** R₈ = 3280 = 11111111₃ — certificate expiry circle */
  REPUNIT_R8: 3280,
  /** R₉ = 9841 = 111111111₃ — archival circle */
  REPUNIT_R9: 9841,

  /** Checksum modulus: R₆ = 364 */
  REPUNIT_CHECKSUM_MODULUS: 364,
  /** Checksum trit count: 6 (digits in R₆ base-3) */
  REPUNIT_CHECKSUM_TRITS: 6,
} as const;

/**
 * Calendar equivalents for repunit circles (precomputed).
 *
 * These are approximate — the exact mapping depends on which DOTs
 * fall within the span relative to the Salvi Epoch start date.
 */
export const REPUNIT_CALENDAR_EQUIVALENTS = {
  R6: { circleDays: 364,  calendarDays: 365,  approxYears: 1 },
  R7: { circleDays: 1093, calendarDays: 1096, approxYears: 3 },   // floor(1093/364) = 3 DOTs
  R8: { circleDays: 3280, calendarDays: 3289, approxYears: 9 },   // 9 DOTs
  R9: { circleDays: 9841, calendarDays: 9868, approxYears: 27 },  // 27 DOTs
} as const;

// ============================================================
// USAGE CONTEXTS
// ============================================================

/**
 * Recommended repunit circle assignments for framework operations.
 *
 * These are geometric intervals — not calendar periods.
 * TSA and HPTP should reference circleDays, with calendar
 * equivalents computed on demand via circleDaysToCalendarDays().
 */
export const REPUNIT_USAGE = {
  /** Daily agent scheduling cycle */
  AGENT_MICRO: 'R3',     // 13 circle-days
  /** Key rotation interval (CNSA 2.0 lifecycle) */
  KEY_ROTATION: 'R7',    // 1093 circle-days
  /** Certificate expiry interval */
  CERT_EXPIRY: 'R8',     // 3280 circle-days
  /** Archival timestamp interval */
  ARCHIVAL: 'R9',        // 9841 circle-days
} as const;

File: ternary-math/src/repunit_circles.rs
// Copyright (c) 2025-2026 Capomastro Holdings Ltd. (Canada)
// Patent(s) Pending — All Rights Reserved — Applied Physics Division
//
// Repunit Circle Constants — Rust Kernel Mirror
// Mirrors: shared/repunit-circles.ts
//
// Base-3 repunits R(n) = (3^n - 1) / 2 define geometric cycle lengths.
// These are circle-days (pure geometry), NOT calendar days.
// Calendar conversion requires DOT (Day Out of Time) insertion.

/// Compute base-3 repunit: R(n) = (3^n - 1) / 2
pub const fn repunit(n: u32) -> u64 {
    let mut pow3: u64 = 1;
    let mut i = 0;
    while i < n {
        pow3 *= 3;
        i += 1;
    }
    (pow3 - 1) / 2
}

// Named constants — derived from constants.rs source of truth.
pub const REPUNIT_R3: u64 = crate::constants::T_REPUNIT_3.host_u32() as u64;
pub const REPUNIT_R4: u64 = crate::constants::T_REPUNIT_4.host_u32() as u64;
pub const REPUNIT_R5: u64 = crate::constants::T_REPUNIT_5.host_u32() as u64;
pub const REPUNIT_R6: u64 = crate::constants::T_REPUNIT_6.host_u32() as u64;
pub const REPUNIT_R7: u64 = crate::trit_int::TritInt::repunit(7).host_u32() as u64;
pub const REPUNIT_R8: u64 = crate::trit_int::TritInt::repunit(8).host_u32() as u64;
pub const REPUNIT_R9: u64 = crate::trit_int::TritInt::repunit(9).host_u32() as u64;

/// Full ternary circle in circle-days.
pub const FULL_CIRCLE_DAYS: u64 = REPUNIT_R6; // 364

/// Convert circle-days to calendar days (DOT-aware).
/// Each complete 364-day rotation inserts 1 DOT.
pub const fn circle_days_to_calendar_days(circle_days: u64) -> u64 {
    let completed_circles = circle_days / FULL_CIRCLE_DAYS;
    circle_days + completed_circles
}

/// Repunit factorization: R(2n) = R(n) × (3^n + 1)
pub const fn repunit_factorization(n: u32) -> (u64, u64, u64) {
    let rn = repunit(n);
    let mut pow3: u64 = 1;
    let mut i = 0;
    while i < n {
        pow3 *= 3;
        i += 1;
    }
    let cofactor = pow3 + 1;
    (rn * cofactor, rn, cofactor)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_repunit_values() {
        assert_eq!(repunit(1), 1);
        assert_eq!(repunit(2), 4);
        assert_eq!(repunit(3), REPUNIT_R3);
        assert_eq!(repunit(4), REPUNIT_R4);
        assert_eq!(repunit(5), REPUNIT_R5);
        assert_eq!(repunit(6), REPUNIT_R6);
        assert_eq!(repunit(7), REPUNIT_R7);
        assert_eq!(repunit(8), REPUNIT_R8);
        assert_eq!(repunit(9), REPUNIT_R9);
    }

    #[test]
    fn test_full_circle_is_r6() {
        assert_eq!(FULL_CIRCLE_DAYS, 364);
        assert_eq!(REPUNIT_R6, 364);
    }

    #[test]
    fn test_circle_to_calendar_single() {
        // 1 circle = 364 circle-days → 365 calendar days (364 + 1 DOT)
        assert_eq!(circle_days_to_calendar_days(364), 365);
    }

    #[test]
    fn test_circle_to_calendar_partial() {
        // Less than 1 full circle → no DOT insertion
        assert_eq!(circle_days_to_calendar_days(100), 100);
        assert_eq!(circle_days_to_calendar_days(363), 363);
    }

    #[test]
    fn test_circle_to_calendar_r7() {
        // R₇ = 1093 circle-days, floor(1093/364) = 3 complete circles → 3 DOTs
        assert_eq!(circle_days_to_calendar_days(REPUNIT_R7), 1096);
    }

    #[test]
    fn test_circle_to_calendar_r8() {
        // R₈ = 3280, floor(3280/364) = 9 complete circles → 9 DOTs
        assert_eq!(circle_days_to_calendar_days(REPUNIT_R8), 3289);
    }

    #[test]
    fn test_circle_to_calendar_r9() {
        // R₉ = 9841, floor(9841/364) = 27 complete circles → 27 DOTs
        assert_eq!(circle_days_to_calendar_days(REPUNIT_R9), 9868);
    }

    #[test]
    fn test_factorization_identity() {
        // R(2n) = R(n) × (3^n + 1)
        for n in 1..=4 {
            let (r2n, rn, cofactor) = repunit_factorization(n);
            assert_eq!(r2n, repunit(2 * n), "R(2×{}) factorization failed", n);
            assert_eq!(r2n, rn * cofactor);
        }
    }

    #[test]
    fn test_powers_of_3_in_calendar_approximations() {
        // R₆ → ~1 year (3⁰), R₇ → ~3 years (3¹), R₈ → ~9 years (3²), R₉ → ~27 years (3³)
        // These are approximate but the powers-of-3 pattern is structurally inherent.
        let r6_years = circle_days_to_calendar_days(REPUNIT_R6) as f64 / 365.0;
        let r7_years = circle_days_to_calendar_days(REPUNIT_R7) as f64 / 365.0;
        let r8_years = circle_days_to_calendar_days(REPUNIT_R8) as f64 / 365.0;
        let r9_years = circle_days_to_calendar_days(REPUNIT_R9) as f64 / 365.0;

        assert!((r6_years - 1.0).abs() < 0.01);
        assert!((r7_years - 3.0).abs() < 0.01);
        assert!((r8_years - 9.0).abs() < 0.02);
        assert!((r9_years - 27.0).abs() < 0.05);
    }
}
